src/joyride/extension.cljs

```diff
@@ -16,10 +16,10 @@
     (.push (.-subscriptions context) disposable)))
 
 (defn- clear-disposables! []
-  (swap! db/!app-db assoc :disposables [])
-  (p/run! (fn [^js disposable]
+  (run! (fn [^js disposable]

```

Comment:
Might we risk starting to add disposable before we've cleared the old ones if we don't wait for the promises to resolve? It won't probably matter in practice, but there's a small risk of two instances of the same subscription being active at the same time, isn't it?

Reply:
The API docs repeatedly warn "Note that an asynchronous function is not awaited." when talking about disposables.

The (obvious, yet surprising) translation is that calling `dispose` on an instance of `Disposable` does not return a promise, even if the underlying dispose function does. Not sure about the mechanics or reason, but this is pretty simple to demonstrate in the repl.

```clojure
cljs꞉joyride.extension꞉>
(let [disposable (vscode/Disposable.
                    (fn []
                      (println "disposing")
                       ;; underlying disposal function returns a promise
                      (p/delay 1 "disposed")))]
    (.dispose disposable))

 ;; The promise-returning function is called
disposing
 ;; but `.dispose` immediately returns nil instead of the promise.
nil 
 ;; The result of the promise, and any chance to wait for resolution is lost.
```
Wrapping the disposal process in `p/run!` here simply wraps a bunch of nils in a promise, rather than actually awaiting any existing promises. We get a false impression of safety. If a race somehow does exist, there really is nothing we can do now.

Shouldn't be a problem though because separate instances are just that, "separate instances". JavaScript does not have meaningful equality semantics. Outside of the weird type coercion, JS generally relies on strict referential equality.
```javascript
const o = {a: 1}
o === o            // ==> true
o === {a: 1}       // ==> false
```
So the only way to get race conditions here is from strict sharing of underlying resources. Presumably by this point in time (having chosen to implement such a no-wait disposal process) Microsoft will have worked out any possible race conditions from Disposables they issue? 🤞

The only time we should need to worry about races is if we generate our own disposables, and connect to existing resources, but then terminate those resources in `dispose` as if we own them. (Which, of course, we should generally avoid.) For example:
```clojure
(let [repl (if (running-repl?)
              ;; try to reuse existing repl
             (connect-to-repl)
              ;; else create our own
             (start-and-connect-to-repl))
      disposable #js{dispose 
                      ;; blindly kill the repl, even if another party created it
                     (fn [kill repl])}]
  .push (.-subscriptions context) disposable)
```

Reply:
Thanks for clarifying this! 🙏